inference/server/oasst_inference_server/routes/workers.py

```diff
@@ -85,191 +72,321 @@ def get_work_request_container(work_request_map: WorkRequestContainerMap, reques
+    except chat_schema.MessageTimeoutException as e:
+        logger.warning(f"Message timed out before work could be initiated: {e.message.id=}")
+        await handle_timeout(message=e.message)
+    finally:
+        futures.ensure_listening_for_work_requests()
+
+
+async def _handle_worker_response(
+    worker_response: inference.WorkerResponse,
+    context: HandleWorkerContext,
+    futures: FuturesManager,
+    session: SessionManager,
+):
+    try:
+        match worker_response.response_type:
+            case "pong":
+                worker_response = cast(inference.PongResponse, worker_response)
+                await session.update(worker_response.metrics)
+            case "token":
+                worker_response = cast(inference.TokenResponse, worker_response)
+                await handle_token_response(
+                    work_request_map=context.work_request_map,
+                    response=worker_response,
+                )
+            case "generated_text":
+                worker_response = cast(inference.GeneratedTextResponse, worker_response)
+                await handle_generated_text_response(
+                    work_request_map=context.work_request_map,
+                    response=worker_response,
+                )
+                await session.update(worker_response.metrics)
+            case "error":
+                worker_response = cast(inference.ErrorResponse, worker_response)
+                await handle_error_response(
+                    work_request_map=context.work_request_map,
+                    response=worker_response,
+                )
+                await session.update(worker_response.metrics)
+            case "general_error":
+                worker_response = cast(inference.GeneralErrorResponse, worker_response)
+                await handle_general_error_response(
+                    response=worker_response,
+                )
+                await session.update(worker_response.metrics)
+            case "safe_prompt":
+                logger.info("Received safe prompt response")
+                worker_response = cast(inference.SafePromptResponse, worker_response)
+                await handle_safe_prompt_response(
+                    response=worker_response,
+                    work_request_map=context.work_request_map,
+                )
+            case "plugin_intermediate":
+                worker_response = cast(inference.PluginIntermediateResponse, worker_response)
+                await handle_plugin_intermediate_response(
+                    work_request_map=context.work_request_map,
+                    response=worker_response,
+                )
+            case _:
+                raise RuntimeError(f"Unknown response type: {worker_response.response_type}")
+    finally:
+        if len(futures._futures) == 0:
+            futures.ensure_listening_for_work_requests()
+        futures.ensure_listening_to_worker_responses()
```

Comment:
In order to avoid logical changes as much as possible, I kept this piece of code here as it was, but that now means that it accesses the internal structure of the `FuturesManager`.

I'm considering whether it might be nicer to not throw all the futures into a single set, but instead keep one future in a specific variable for listening to worker responses, and a set for listening to the potentially multiple futures with which we listen for work requests. The futures manager could then have only a single method that ensures that there's at least one pending future of each type, and only that would be called from the outside at the end of each loop